### PR TITLE
234: Relay protocol

### PR DESCRIPTION
**Pull request by qdddddd**
In this PR I implemented our probabilistic relay model, that a block is relayed with probability 1 if it's `countDown` in NetMessage header is greater than 0, and is relayed with probability less than 1 if the `countDown` becomes 0.

### CODE REVIEW

File: src/messages/net_message.h [hunk 1]
```diff
@@ -38,21 +38,30 @@ class NetMessage {
         NONE,
     };
 
-    explicit NetMessage(NetMessage::Type type = NONE) : type_(type) {}
+    explicit NetMessage(NetMessage::Type type = NONE, uint8_t countDown = 255) : type_(type), countDown_(countDown) {}
 
     virtual ~NetMessage() {}
 
-    static unique_message_t MessageFactory(uint32_t type, VStream& s);
+    static unique_message_t MessageFactory(uint8_t type, uint8_t coundDown, VStream& s);
```
Comment (bassland) [inline src/messages/net_message.h]:
```suggestion
    static unique_message_t MessageFactory(uint8_t type, uint8_t countDown, VStream& s);
```

File: src/peer/peer_manager.cpp [hunk 2]
```diff
@@ -468,25 +491,39 @@ void PeerManager::RelayAddressMsg(AddressMessage& message, const PeerPtr& msg_fr
...
 }
 
+std::vector<PeerPtr> PeerManager::RandomlySelect(size_t size) {
+    std::shared_lock<std::shared_mutex> lk(peerLock_);
+
+    assert(peerMap_.size() == peerMapIndices_.size());
+    size = std::min(size, peerMapIndices_.size());
+
+    std::vector<PeerPtr> result;
+    auto indices = peerMapIndices_;
```
Comment (bassland) [inline src/peer/peer_manager.cpp]: I suggest generating indices set from peerMap in the function

File: src/utils/concurrent_container.h [hunk 3]
```diff
@@ -14,6 +14,9 @@
 #define READER_LOCK(mu) std::shared_lock<std::shared_mutex> reader(mu);
 #define WRITER_LOCK(mu) std::unique_lock<std::shared_mutex> writer(mu);
 
+auto key_selector   = [](const auto& pair) { return pair.first; };
```
Comment (0ersted) [inline src/utils/concurrent_container.h]: why it is moved from private to outside?

**Comment by qdddddd** [comment]
Because it is used outside.

Sent from my iPhone. Please excuse any tpyos.

On Jan 21, 2020, at 9:06 PM, 0ersted <notifications@github.com> wrote:

﻿

@0ersted commented on this pull request.

________________________________

In src/utils/concurrent_container.h<https://github.com/EPI-ONE/epic/pull/175#discussion_r368988418>:

> @@ -14,6 +14,9 @@
 #define READER_LOCK(mu) std::shared_lock<std::shared_mutex> reader(mu);
 #define WRITER_LOCK(mu) std::unique_lock<std::shared_mutex> writer(mu);

+auto key_selector   = [](const auto& pair) { return pair.first; };


why it is moved from private to outside?

File: src/utils/concurrent_container.h (hunk 3 above)
Comment (qdddddd) [reply src/utils/concurrent_container.h]: Because it’s paired with the `value_selector`

Comment (0ersted) [reply src/utils/concurrent_container.h]: I would recommend not moving them out but writing this simple lambda function explicitly when using for encapsulation and readability.

**Pull request merged**